### PR TITLE
Provide CustomNSError default implementation

### PR DESCRIPTION
**Pull request by zayass**
Pull default CustomNSError implementation from https://github.com/apple/swift/blob/master/stdlib/public/SDK/Foundation/NSError.swift

### CODE REVIEW

**Comment by ianpartridge** [comment]
Can we add some [tests](https://github.com/apple/swift-corelibs-foundation/blob/master/TestFoundation/TestNSError.swift) in this area?

**Comment by zayass** [comment]
@ianpartridge I add some tests but have no environment to properly run it. Can you trigger CI build?

File: Foundation/NSError.swift [hunk 1]
```diff
@@ -274,6 +274,47 @@ public protocol CustomNSError : Error {
...
+extension CustomNSError where Self: RawRepresentable, Self.RawValue: SignedInteger {
+    // The error code of Error with integral raw values is the raw value.
+    public var errorCode: Int {
+        return numericCast(self.rawValue)
+    }
+    
+    // conflict resolution
+	var _code: Int {
+		return errorCode
+	}
```
Comment (xwu) [inline Foundation/NSError.swift]: Can you fix the indenting on these lines?

File: Foundation/NSError.swift [hunk 2]
```diff
@@ -274,6 +274,47 @@ public protocol CustomNSError : Error {
...
+    }
+}
+
+extension CustomNSError where Self: RawRepresentable, Self.RawValue: SignedInteger {
+    // The error code of Error with integral raw values is the raw value.
+    public var errorCode: Int {
+        return numericCast(self.rawValue)
+    }
+    
+    // conflict resolution
```
Comment (xwu) [inline Foundation/NSError.swift]: What does this comment mean?

File: Foundation/NSError.swift [hunk 3]
```diff
@@ -274,6 +274,47 @@ public protocol CustomNSError : Error {
...
+extension CustomNSError where Self: RawRepresentable, Self.RawValue: UnsignedInteger {
+    // The error code of Error with integral raw values is the raw value.
+    public var errorCode: Int {
+        return numericCast(self.rawValue)
+    }
+    
+    // conflict resolution
+	var _code: Int {
+		return errorCode
+	}
```
Comment (xwu) [inline Foundation/NSError.swift]: Also here.

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by ianpartridge** [comment]
@zayass 

```
TestFoundation/TestNSError.swift:75:14: error: type 'SwiftError' does not conform to protocol 'Error'
        enum SwiftError : Int, Error, CustomNSError {
             ^
Swift.Error:101:16: note: multiple matching properties named '_code' with type 'Int'
    public var _code: Int { get }
               ^
Swift.Error:2:16: note: candidate exactly matches
    public var _code: Int { get }
               ^
Swift.Error:2:16: note: candidate exactly matches
    public var _code: Int { get }
               ^
Foundation.Error:5:16: note: candidate exactly matches
    public var _code: Int { get }
               ^
TestFoundation/TestNSError.swift:85:14: error: type 'SwiftError' does not conform to protocol 'Error'
        enum SwiftError : UInt, Error, CustomNSError {
             ^
Swift.Error:101:16: note: multiple matching properties named '_code' with type 'Int'
    public var _code: Int { get }
               ^
Swift.Error:2:16: note: candidate exactly matches
    public var _code: Int { get }
               ^
Swift.Error:2:16: note: candidate exactly matches
    public var _code: Int { get }
               ^
Foundation.Error:5:16: note: candidate exactly matches
    public var _code: Int { get }
               ^
```

**Comment by zayass** [comment]
@ianpartridge fix it. Just forget `public` for `_code` extesnion

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by ianpartridge** [comment]
@swift-ci please test

**Comment by alblue** [comment]
Looks good to me. WDYT @ianpartridge ?

File: TestFoundation/TestNSError.swift [hunk 4]
```diff
@@ -40,4 +47,45 @@ class TestNSError : XCTestCase {
...
         XCTAssertEqual(error.localizedDescription, "Localized!")
     }
+
+    func test_CustomNSError_domain() {
+        XCTAssertEqual(SwiftCustomNSError.errorDomain, "TestFoundation.SwiftCustomNSError")
+    }
+
+    func test_CustomNSError_userInfo() {
+        let userInfo = SwiftCustomNSError().errorUserInfo
+            XCTAssertTrue(userInfo.isEmpty)
```
Comment (ianpartridge) [inline TestFoundation/TestNSError.swift]: Can you fix the indentation of this line?

File: TestFoundation/TestNSError.swift [hunk 5]
```diff
@@ -40,4 +47,45 @@ class TestNSError : XCTestCase {
...
+
+        XCTAssertEqual(SwiftError.zero.errorCode, 0)
+        XCTAssertEqual(SwiftError.one.errorCode,  1)
+        XCTAssertEqual(SwiftError.two.errorCode,  2)
+    }
+
+    func test_CustomNSError_errorCodeRawInt() {
+        enum SwiftError : Int, Error, CustomNSError {
+            case minusOne  = -1
+            case fourtyTwo = 42
```
Comment (ianpartridge) [inline TestFoundation/TestNSError.swift]: Spelling: `fortyTwo` 🙂

File: TestFoundation/TestNSError.swift [hunk 6]
```diff
@@ -40,4 +47,45 @@ class TestNSError : XCTestCase {
...
+        }
+
+        XCTAssertEqual(SwiftError.minusOne.errorCode,  -1)
+        XCTAssertEqual(SwiftError.fourtyTwo.errorCode, 42)
+    }
+
+    func test_CustomNSError_errorCodeRawUInt() {
+        enum SwiftError : UInt, Error, CustomNSError {
+            case fourtyTwo = 42
```
Comment (ianpartridge) [inline TestFoundation/TestNSError.swift]: Spelling: `fortyTwo` 🙂

**Comment by ianpartridge** [comment]
Minor tweaks but otherwise this looks good to me 👍

**Comment by zayass** [comment]
@ianpartridge fixed

**Comment by ianpartridge** [comment]
@swift-ci please test and merge

**Comment by ianpartridge** [comment]
@swift-ci please test and merge

**Comment by ianpartridge** [comment]
Thanks @zayass 👍

**Comment by zayass** [comment]
May be I can cherry pick this to 3.1 branch. or it is closed?

**Comment by ianpartridge** [comment]
There isn't going to be another release from the 3.1 branch, I'm afraid.

**Pull request merged**